### ISSUE TITLE
False positive when using $this in will()

### ISSUE THREAD

**Issue by dontub** [Jan0707/phpstan-prophecy#39]
Code like
```
$userProphecy->setName('foo')->will(function () {
    $this->getName()->willReturn('foo');
});
```
leads to a false positive:
> Call to an undefined method SomeTest::getName().

**Comment by Jean85** [comment]
I believe that that is correct. In that context, `$this` is the test class in which you're writing the prophecy, not the prophecy itself.

**Comment by dontub** [comment]
When the callback is executed, `$this` is the prophecy. You can also find this approach in the Prophecy documentation: https://github.com/phpspec/prophecy#method-prophecies-idempotency
This would be possible instead:
```php
$userProphecy->setName('foo')->will(function () use ($userProphecy) {
    $userProphecy->getName()->willReturn('foo');
});
```
I could use this approach everywhere to avoid the false positive. Though because using `$this` is recommended by the Prophecy documentation, it should be fixed IMHO.

**Comment by Jean85** [comment]
I'm not sure this is a "bug" of this library, it's more generic than that, since it's related to callables bindings.

**Comment by Xerkus** [comment]
Normally `$this` is bound at the closure creation time, unless it is explicitly re-binded later.

Prophecy does exactly that in [CallbackPromise](https://github.com/phpspec/prophecy/blob/7e272180527c34a97680de85eb5aba0847a664e0/src/Prophecy/Promise/CallbackPromise.php#L61) and [CallbackPrediction](https://github.com/phpspec/prophecy/blob/7e272180527c34a97680de85eb5aba0847a664e0/src/Prophecy/Prediction/CallbackPrediction.php#L60)

@Jean85 I would argue this case should be handled by this library, if it can be done in practical way.

**Comment by Jean85** [comment]
That's exactly the issue here. Static analysis cannot know if the closure is re-binded later, so it cannot know if that's wrong or not. Maybe we can report the bug to PHPStan, since this could be a false positive.

**Comment by Xerkus** [comment]
rebinding closure passed as an argument is a part of prophecy api here. No static analysis can know it. It is definitely not an issue to be propagated to PHPStan itself.

**Comment by Jan0707** [comment]
I feel we need input from @ondrejmirtes here. Ping!

**Comment by ondrejmirtes** [comment]
There's an open issue about this: https://github.com/phpstan/phpstan/issues/149

**Comment by Jan0707** [comment]
I will close this for now. We can re-discuss once phpstan/phpstan#149 has been dealt with. Thanks for the input @ondrejmirtes !
